fv3gfs/util/halo_data_transformer.py

```diff
@@ -0,0 +1,880 @@
+        # Thus we rotate that number of times counterclockwise before sending,
+        # to get the right final orientation. We apply those rotations to the
+        # indices here to prepare for a straightforward copy in cu kernel
+        arr_indices = rotate_scalar_data(arr_indices, dims, cp, -rotation)
+    arr_indices_gpu = cp.asarray(arr_indices.flatten(order="C"))
+    INDICES_GPU_CACHE[key] = arr_indices_gpu
+
+
+# ------------------------------------------------------------------------
+# HaloDataTransformer helpers
+
+
+def _slices_length(slices: Tuple[slice]) -> int:
+    """Compute linear size from slices."""
+    length = 1
+    for s in slices:
+        assert s.step is None
+        length *= abs(s.start - s.stop)
+    return length
+
+
+@dataclass
+class HaloExchangeData:
+    """Memory description of the data exchanged.
+    
+    Args:
+        specification: memory layout of the data
+        pack_slices: indexing to pack
+        pack_clockwise_rotation: required data rotation when packing
+        unpack_slices: indexing to unpack
+    """
+
+    specification: HaloUpdateSpec
+    pack_slices: Tuple[slice]
+    pack_clockwise_rotation: int
+    unpack_slices: Tuple[slice]
+
+    def __post_init__(self):
+        self._id = uuid1()
+        self._pack_buffer_size = _slices_length(self.pack_slices)
+        self._unpack_buffer_size = _slices_length(self.unpack_slices)
+
+
+class _HaloDataTransformerType(Enum):
+    """Dimensionality of the data in the packed buffer."""
+
+    UNKNOWN = 0
+    SCALAR = 1
+    VECTOR = 2
+
+
+# ------------------------------------------------------------------------
+# HaloDataTransformer classes
+
+
+class HaloDataTransformer:
+    """Transform data to exchange in a format optimized for network communication.
+
+    Current strategy: pack/unpack multiple nD array into/from a single buffer.
+    Offers a send and a recv buffer to use for exchanging data.
+
+    The class is responsible for packing & unpacking, not communication.
+    Order of operations:
+    - get HaloDataTransformer via get_from_numpy_module
+    (user should re-use the same if it goes to the same destination)
+    - add N transformation with the proper halo specification via add_scalar_specification or
+     add_vector_specification (can't mix and match scalar & vector)
+    - compile the transformer via compile()
+    [From here get_unpack_buffer() is valid]
+    - async_pack
+    - synchronize
+    [From here get_pack_buffer() is valid]
+    [... user should communicate the buffers...]
+    - async_unpack
+    - finalize
+    [All buffers are returned to their buffer cache, therefore invalid]
+
+    If the user doesn't want the buffer return to the cache (optimal behavior if the same exchange
+    is to be done multiple times), it can use `synchronize` instead of `finalize`
+    """
+
+    _pack_buffer: Optional[Buffer]
+    _unpack_buffer: Optional[Buffer]
```

Comment:
Moving the buffers out would allow for a less state-full class but then it would be responsibility of the user of the class to give the same buffer. I am not sure that makes the usage of this class easier really